pkg/stores/tournament/cache.go

```diff
@@ -0,0 +1,344 @@
+	uncachedTournament, err := c.backing.Get(ctx, id)
+	if err == nil {
+		c.cache.Add(id, uncachedTournament)
+	}
+	return uncachedTournament, err
+}
+
+// Set sets a tournament in the cache, AND in the backing store. This ensures if the
+// node crashes the tournament doesn't just vanish.
+func (c *Cache) Set(ctx context.Context, tm *entity.Tournament) error {
+	return c.setOrCreate(ctx, tm, false)
+}
+
+// Create creates the tournament in the cache as well as the store.
+func (c *Cache) Create(ctx context.Context, tm *entity.Tournament) error {
+	return c.setOrCreate(ctx, tm, true)
+}
+
+func (c *Cache) setOrCreate(ctx context.Context, tm *entity.Tournament, isNew bool) error {
+	var err error
+	if isNew {
+		err = c.backing.Create(ctx, tm)
+	} else {
+		err = c.backing.Set(ctx, tm)
+	}
+	if err != nil {
+		return err
+	}
+	c.cache.Add(tm.UUID, tm)
+	return nil
+}
+
+func (c *Cache) SetTournamentControls(ctx context.Context, id string, name string, description string, controls *entity.TournamentControls) error {
+
+	t, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	if t.IsStarted {
+		return errors.New("Cannot change tournament controls after it has started.")
+	}
+
+	t.Name = name
+	t.Description = description
+	t.Controls = controls
+
+	return c.backing.Set(ctx, t)
+}
+
+func (c *Cache) AddDirectors(ctx context.Context, id string, directors *entity.TournamentPersons) error {
+
+	t, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	err = entity.AddDirectors(t, directors)
+	if err != nil {
+		return err
+	}
+
+	return c.backing.Set(ctx, t)
+}
+
+func (c *Cache) RemoveDirectors(ctx context.Context, id string, directors *entity.TournamentPersons) error {
+
+	t, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	err = entity.RemoveDirectors(t, directors)
+	if err != nil {
+		return err
+	}
+
+	return c.backing.Set(ctx, t)
+}
+
+func (c *Cache) AddPlayers(ctx context.Context, id string, players *entity.TournamentPersons) error {
+
+	t, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	err = entity.AddPlayers(t, players)
+	if err != nil {
+		return err
+	}
+
+	return c.backing.Set(ctx, t)
+}
+
+func (c *Cache) RemovePlayers(ctx context.Context, id string, players *entity.TournamentPersons) error {
+
+	t, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	err = entity.RemovePlayers(t, players)
+	if err != nil {
+		return err
+	}
+
+	return c.backing.Set(ctx, t)
+}
+
+func (c *Cache) SetPairing(ctx context.Context, id string, playerOneId string, playerTwoId string, round int) error {
+
+	t, err := c.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+
+	if !t.IsStarted {
+		return errors.New("Cannot set tournament pairings before the tournament has started.")
+	}
+
+	err = t.TournamentManager.SetPairing(playerOneId, playerTwoId, round)
+	if err != nil {
+		return err
+	}
+
+	return c.backing.Set(ctx, t)
+}
+
+func (c *Cache) SetResult(ctx context.Context,
```

Comment:
These functions should be moved to the tournament package; cache should only use Set and Get (and maybe a couple more functions)

Reply:
DOne.